Tests/test-dataframe.lisp

```diff
@@ -0,0 +1,554 @@
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-valid-col-slice df))
+                       (funcall (gen-valid-function 'map)))))))
+    (finishes (map-data-frame df row-slice col-slice function))))
+
+(def-test map-data-frame.invalid ()
+  (for-all (((df row-slice col-slice row-slice* col-slice*)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-valid-col-slice df))
+                       (funcall (gen-invalid-row-slice df))
+                       (funcall (gen-invalid-col-slice df))))))
+            (function (gen-valid-function 'map))
+            (function* (gen-invalid-function 'map)
+                       (if (or (and (arrayp row-slice)
+                                    (zerop (array-total-size row-slice)))
+                               (and (arrayp col-slice)
+                                    (zerop (array-total-size col-slice))))
+                           nil
+                           t)))
+    (signals error (map-data-frame df row-slice col-slice function*))
+    (signals <invalid-slice> (map-data-frame df row-slice col-slice* function))
+    (signals <invalid-slice> (map-data-frame df row-slice* col-slice function))
+    (signals <invalid-slice> (map-data-frame df row-slice* col-slice* function))
+    (signals <invalid-slice> (map-data-frame df row-slice* col-slice* function*))))
+
+(def-test map-data-frame-rows ()
+  (for-all (((df row-slice function)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-valid-function 'row)))))))
+    (finishes (map-data-frame-rows df row-slice function))))
+
+(def-test map-data-frame-rows.invalid ()
+  (for-all (((df row-slice row-slice*)
+             (lambda ()
+               (let ((df (funcall
+                          (gen-data-frame (funcall (gen-integer :min 1 :max +max-rows+))
+                                          (funcall (gen-integer :min 1 :max +max-cols+))))))
+                 (list df
+                       (funcall (gen-valid-row-slice df))
+                       (funcall (gen-invalid-row-slice df))))))
+            (function (gen-valid-function 'row))
+            (function* (gen-invalid-function 'row) 
+                       (if (and (arrayp row-slice)
+                                (zerop (array-total-size row-slice)))
+                           nil
+                           t)))
```

Comment:
replace if with just a condition:
`(not (and (arrayp …) (zerop …))`

Reply:
Good style; thanks.